### PR TITLE
Add serialization/deserialization support for designated protocols fo…

### PR DESCRIPTION
**Pull request by rudkx**
…r operators.

Parsing for these protocols in operator declarations is gated by
-enable-operator-designated-protocols (added in
https://github.com/apple/swift/pull/19145).

### CODE REVIEW

File: include/swift/Serialization/ModuleFormat.h [hunk 1]
```diff
@@ -1089,8 +1089,9 @@ namespace decls_block {
   template <unsigned Code>
   using UnaryOperatorLayout = BCRecordLayout<
     Code, // ID field
-    IdentifierIDField, // name
-    DeclContextIDField // context decl
+    IdentifierIDField,  // name
+    DeclContextIDField, // context decl
+    TypeIDField         // protocol
```
Comment (jrose-apple) [inline include/swift/Serialization/ModuleFormat.h]: This should really be serialized as a Decl, not a Type, no?

File: lib/Serialization/Deserialization.cpp [hunk 2]
```diff
@@ -3376,37 +3376,58 @@ ModuleFile::getDeclCheckedImpl(DeclID DID) {
...
 
-    decls_block::PrefixOperatorLayout::readRecord(scratch, nameID,
-                                                  contextID);
+    decls_block::PrefixOperatorLayout::readRecord(scratch, nameID, contextID,
+                                                  protoID);
     auto DC = getDeclContext(contextID);
-    declOrOffset = createDecl<PrefixOperatorDecl>(DC, SourceLoc(),
-                                                  getIdentifier(nameID),
-                                                  SourceLoc());
+    TypeLoc protoTyLoc = TypeLoc::withoutLoc(getType(protoID));
```
Comment (jrose-apple) [inline lib/Serialization/Deserialization.cpp]: Note that this has poor recovery (read: will crash) if the protocol can't be deserialized for whatever reason. That's probably fine for now, but really there should be a recovery mechanism.

File: lib/Serialization/Deserialization.cpp [hunk 3]
```diff
@@ -3376,37 +3376,58 @@ ModuleFile::getDeclCheckedImpl(DeclID DID) {
...
+    Identifier protocolName;
+    if (auto protoTy = protoTyLoc.getType())
+      if (auto *protoDecl = protoTy->getNominalOrBoundGenericNominal())
+        protocolName = protoDecl->getName();
```
Comment (jrose-apple) [inline lib/Serialization/Deserialization.cpp]: I only skimmed your previous proposal, but this confuses me: an identifier isn't enough to uniquely identify a type, so why bother?

File: include/swift/Serialization/ModuleFormat.h (hunk 1 above)
Comment (rudkx) [reply include/swift/Serialization/ModuleFormat.h]: Yes that makes sense given that I am only using the type to get to the decl at this point. An earlier version of this added constructors for OperatorDecls that took TypeLocs.

File: lib/Serialization/Deserialization.cpp (hunk 2 above)
Comment (rudkx) [reply lib/Serialization/Deserialization.cpp]: What's a better approach? We seem to be doing something very similar in other places.

Comment (jrose-apple) [reply lib/Serialization/Deserialization.cpp]: Look for the uses of `getTypeChecked` elsewhere in the file. Moving over to that for everything is a big task, but we probably want new fields to be thinking about recovery.

**Comment by rudkx** [comment]
I'll push another version of this once https://github.com/apple/swift/pull/19202 lands.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Pull request merged**